music/src/core/data.ts

```diff
@@ -396,6 +409,143 @@ export class MelodyConverter extends DataConverter {
+ * `toNoteSequence` method is left undefined here but can be used by subclasses
+ * for debugging purposes.
+ *
+ * @param numSteps The length of each sequence.
+ * @param minPitch The minimum melody pitch.
+ * @param maxPitch The maximum melody pitch.
+ * @param ignorePolpyhony Whether to raise an error if multiple notes start at
+ * the same step.
+ * @param numSegments (Optional) The number of conductor segments.
+ */
+abstract class MelodyControlConverter extends DataConverter {
+  readonly minPitch: number;
+  readonly maxPitch: number;
+  readonly ignorePolyphony: boolean;
+  readonly melodyControl: MelodyControl;
+  readonly depth: number;
+  readonly endTensor: tf.Tensor1D;
+
+  constructor(args: MelodyConverterArgs, melodyControl: MelodyControl) {
+    super(args);
+    this.minPitch = args.minPitch;
+    this.maxPitch = args.maxPitch;
+    this.ignorePolyphony = args.ignorePolyphony;
+    this.melodyControl = melodyControl;
+    this.depth = melodyControl.depth;
+  }
+
+  toTensor(noteSequence: INoteSequence): tf.Tensor2D {
+    const melody = Melody.fromNoteSequence(
+        noteSequence, this.minPitch, this.maxPitch, this.ignorePolyphony,
+        this.numSteps);
+    return this.melodyControl.extract(melody);
+  }
+}
+
+/**
+ * Converts between a monophonic, quantized `NoteSequence` containing a melody
+ * and a `Tensor` representing only the *rhythm* of the melody.
+ *
+ * The rhythm is represented as a `numSteps`-by-1 `Tensor` with 1 in the
+ * positions corresponding to steps with a note-on and 0 elsewhere.
+ *
+ * Since the melody cannot be reconstructed from its rhythm alone,
+ * `toNoteSequence` returns a `NoteSequence` with drum hits at the note-on
+ * steps.
+ */
+export interface MelodyRhythmConverterArgs extends BaseConverterArgs {}
+export class MelodyRhythmConverter extends MelodyControlConverter {
+  constructor(args: MelodyConverterArgs) {
+    super(args, new MelodyRhythm());
+  }
+
+  async toNoteSequence(
+      tensor: tf.Tensor2D, stepsPerQuarter?: number, qpm?: number) {
+    // Create a NoteSequence containing the rhythm as drum hits.
+    // This is mainly for debugging purposes.
+    const noteSequence =
+        sequences.createQuantizedNoteSequence(stepsPerQuarter, qpm);
+    const rhythm: Int32Array = await tensor.data() as Int32Array;
+    for (let s = 0; s < rhythm.length; ++s) {
+      if (rhythm[s]) {
+        noteSequence.notes.push(NoteSequence.Note.create({
+          pitch: DEFAULT_DRUM_PITCH_CLASSES[1][0],  // snare
+          quantizedStartStep: s,
+          quantizedEndStep: s + 1,
+          isDrum: true
+        }));
+      }
+    }
+    noteSequence.totalQuantizedSteps = rhythm.length;
+    return noteSequence;
+  }
+}
+
+/**
+ * Converts between a monophonic, quantized `NoteSequence` containing a melody
+ * and a `Tensor` representing only the *shape* of the melody.
+ *
+ * The shape is represented as a `numSteps`-by-3 `Tensor` containing a one-hot
```

Comment:
I think "`[num_steps, 3]`-shaped `Tensor`" is more readable.

Reply:
Done.